packages/overseer-rs/src/monitor.rs

```diff
@@ -0,0 +1,269 @@
+        let lookahead_result = {
+            let mut builder = self.lookahead_builder.lock().await;
+            builder.get_lookahead_data().await
+        };
+
+        let (slot_index_u256, curr_lookahead) = match lookahead_result {
+            Ok((_, data)) => {
+                let LookaheadData {
+                    slotIndex,
+                    currLookahead,
+                    ..
+                } = data;
+                (slotIndex, currLookahead)
+            }
+            Err(err) => {
+                error!(
+                    target: "overseer::service",
+                    error = ?err,
+                    "failed to fetch lookahead data"
+                );
+                return Err(err);
+            }
+        };
+
+        if curr_lookahead.is_empty() {
+            warn!(
+                target: "overseer::service",
+                "lookahead data returned no slots"
+            );
+            return Ok(None);
+        }
+
+        if slot_index_u256 == U256::MAX {
+            warn!(
+                target: "overseer::service",
+                "lookahead slot index is max value; skipping cycle"
+            );
+            return Ok(None);
+        }
+
+        let slot_index: usize = slot_index_u256
+            .try_into()
+            .map_err(|_| anyhow!("lookahead slot index exceeds usize bounds"))?;
+
+        let lookahead_slot = match curr_lookahead.get(slot_index) {
+            Some(slot) => slot,
+            None => {
+                warn!(
+                    target: "overseer::service",
+                    slot_index,
+                    lookahead_len = curr_lookahead.len(),
+                    "lookahead slot index out of bounds"
+                );
+                return Ok(None);
+            }
+        };
+
+        let committer = lookahead_slot.committer;
+        let committer_hex = format!("{committer:#x}");
+        let registration_root = B256::from(lookahead_slot.registrationRoot);
+
+        Ok(Some(crate::types::Preconfirmer {
+            id: committer_hex,
+            registration_root,
+        }))
+    }
+
+    /// Dispatches a blacklist request for the provided violation using the contract client.
+    async fn execute_blacklist(
+        &self,
+        preconfirmer: &crate::types::Preconfirmer,
+        violation: Violation,
+    ) -> Result<()> {
+        match self
+            .blacklist_contract
+            .blacklist(preconfirmer, &violation)
+            .await
+        {
+            Ok(()) => {
+                self.metrics.inc_blacklist_calls();
+                Ok(())
+            }
+            Err(err) => {
+                self.metrics.inc_blacklist_errors();
+                Err(err)
+            }
+        }
+    }
+
+    async fn identify_stalled_transactions(
+        &self,
+        pending: &[PendingTransaction],
+    ) -> Vec<StalledTransaction> {
+        let mut tracker = self.pending_tracker.lock().await;
```

Comment:
method tracks transaction hashes across cycles, but the tracker clears entries for transactions that aren’t seen in the current snapshot immediately. If a transaction is briefly included in a block and then reorged back into the pool, its previous tracking history is lost and its “age” resets. Is this expected ?

Reply:
yeah this is by design as it DID get included at one point so it means the preconfirmer is not censoring